api/src/main/java/org/openmrs/module/insuranceclaims/api/service/fhir/impl/FHIRClaimDiagnosisServiceImpl.java

```diff
@@ -0,0 +1,166 @@
+        }
+
+        return diagnosis;
+    }
+
+    public void setDiagnosisDao(InsuranceClaimDiagnosisDao diagnosisDao) {
+        this.diagnosisDao = diagnosisDao;
+    }
+
+    private void addCodingToDiagnosis(List<Claim.DiagnosisComponent>  diagnosisComponents) {
+        for (Claim.DiagnosisComponent diagnosis: diagnosisComponents) {
+            try {
+                setDiagnosisPrimaryCoding(diagnosis);
+                List<Coding> coding = diagnosis.getDiagnosisCodeableConcept().getCoding();
+                List<CodeableConcept> diagnosisType = coding.stream()
+                        .map(Coding::getSystem)
+                        .map(systemName -> new CodeableConcept().setText(systemName))
+                        .collect(Collectors.toList());
+
+                diagnosis.setType(diagnosisType);
+            } catch (FHIRException exception) {
+                continue;
+            }
+        }
+    }
+    private void setDiagnosisPrimaryCoding(Claim.DiagnosisComponent diagnosis) throws FHIRException {
+        String primaryCoding = InsuranceClaimConstants.PRIMARY_DIAGNOSIS_MAPPING;
+
+        List<Coding> diagnosisCoding = diagnosis.getDiagnosisCodeableConcept().getCoding();
+        for (Coding c: diagnosisCoding) {
+            if (c.getSystem().equals(primaryCoding)) {
+                Collections.swap(diagnosisCoding, 0, diagnosisCoding.indexOf(c));
+                break;
+            }
+        }
+    }
+
+    private void validateDiagnosisCodingSystem(Claim.DiagnosisComponent diagnosis) {
+        //Method assigns diagnosis system (ICD-10, CIEL, etc.) from type to the concept coding if it don't have assigned system
+        if (diagnosis.getType().size() == 0) {
+            return;
+        }
+
+        try {
+            String sys = diagnosis.getType().get(0).getText();
+            CodeableConcept concept = diagnosis.getDiagnosisCodeableConcept();
+            for (Coding coding : concept.getCoding()) {
+                if (coding.getSystem() == null) {
+                    coding.setSystem(sys);
+                }
+            }
+        } catch (FHIRException e) {
+            return;
+        }
+    }
+
+    private static Concept getConceptFromCode(CodeableConcept codeableConcept, List<String> errors) {
+        String conceptCode;
+        String system;
+        Concept concept = null;
+        List<Coding> dts = codeableConcept.getCoding();
+
+        for (Coding coding : dts) {
+            conceptCode = coding.getCode();
+            system = coding.getSystem();
+            if (FHIRConstants.OPENMRS_URI.equals(system)) {
+                concept = Context.getConceptService().getConceptByUuid(conceptCode);
+            } else {
+                String systemName = system;
+                if (systemName != null && !systemName.isEmpty()) {
+                    List<Concept> concepts = Context.getConceptService().getConceptsByMapping(conceptCode, systemName);
+                    if (concepts.size() == 1) {
```

Comment:
why we verify size == 1 ? it is not readable, maybe add new private method or change 1 to constant